Conan update

piponazo opened the pull request:
After a recent change in the **libcurl** conan binaries, our build started to fail. The issues have been reported to the maintainers of that conan package:

https://github.com/bincrafters/community/issues/872

To solve the issue in our CI I just decided to switch to the static version of the libraries to avoid the linking issues with shared libraries. I did not want to spend too much time trying to solve the issue since anyways, each distribution provide different versions of the libraries and those things are already covered by the Gitlab scripts. I just tried to do the minimal amount of work to fix the situation in our travis CI builds. 

After changing to the static version of the library, one of the builders fail with linking errors about the **dl** library:
https://travis-ci.org/Exiv2/exiv2/builds/557689271

That's why I added the second commit.

piponazo committed:
Remove linkage with -ldl

The code using dlopen/dlclose/etc was removed from version.cpp in the past

clanmills commented:
@piponazo If forcing static linking is necessary to restore the CI builders, that's fine.  Let's try to remember to return to shared linking.  Could you open an issue about that for v0.27.3?

I'm looking at the log for the broken build (denial) and see the error.  I'll install a VM this morning and look at that.

```
-- Library rt not found in package, might be system one
-- Library pthread not found in package, might be system one
-- Library expat found /home/travis/conanData/Expat/2.2.6/pix4d/stable/package/ba0d824afa36eebfbb7750b6a5fe285880a17b57/lib/libexpat.a
-- Library ssl found /home/travis/conanData/OpenSSL/1.1.1c/conan/stable/package/0abbb2ea17cdc92f4a2ac8a9e55de717e3b5a9d1/lib/libssl.a
-- Library crypto found /home/travis/conanData/OpenSSL/1.1.1c/conan/stable/package/0abbb2ea17cdc92f4a2ac8a9e55de717e3b5a9d1/lib/libcrypto.a
-- Library dl not found in package, might be system one
-- Library pthread not found in package, might be system one
-- Library z found /home/travis/conanData/zlib/1.2.11/conan/stable/package/76f87539fc90ff313e0b3182641a9bb558a717d2/lib/libz.a
```

I'm building on the build server at the moment.  Last night's build failed because I disabled sshd on rmillsmm-w7 while on vacation.

Do you know if the shared failure in also on xenial or other platforms?

clanmills reviewed (dismissed):
This seems fine.  We'll have to solve the issue on 'xenial' = Ubuntu 16.04.  I already have a VM for that on the buildserver.

clanmills commented:
Build server has built shared libraries, so I can build Exiv2 v0.27.2-RC3 this weekend.

I see the libcurl/shared/linking issue is also on Xenial (Ubuntu 16.04).  I'm investigating.

clanmills commented:
I'm can build and pass the test suite on this branch on Ubuntu 16.04.  Both "pure CMake" and "Conan+CMake" build and pass the test suite.

```
$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
```

I'm surprised to see the file `WORK-IN-PROGRESS` in the code-base as I thought we deleted that before v0.27.0.  I'll remove that in the PR for 0.27.2.3

Perhaps we should `Update Branch` to see if that changes something.  I'm going off-line for a couple of hours to do some garden chores in preparation for Alison's birthday party tomorrow with the family.

D4N commented:
> Perhaps we should `Update Branch` to see if that changes something

Please never press that button, do a `git pull --rebase` instead.

D4N commented:
Something seems to be busted with conan's openSSL package: one of the Travis builds fails due linking issues with openSSL and the appveyor builds are also failing due to issues in the openSSL conan recipe.

clanmills commented:
I never touch that button!  How can the builder perform a 'git pull --rebase'.

Another possibility is that the builder doesn't have to the correct dependancies installed.  I know there's a script for that, however I lost a lot of time on the 0.27 RCs trying unsuccessfully to get that to work.  My build server is set up following the procedure in `README.md`:

```
$ sudo apt install --yes build-essential git clang ccache python3 libxml2-utils cmake python3 libexpat1-dev libz-dev zlib1g-dev libssh-dev libcurl4-openssl-dev libgtest-dev google-mock
```

piponazo commented:
In the Travis-CI builders the idea is to not install the system-packages but bring the dependencies with conan. The Gitlab builders (where we have different linux distributions) bring the dependencies with system-packages. In that way we test both scenarios.

It is weird that know some of the travis builds are failing, I created the PR just after checking that it was passing in the travis-ci/push jobs. 

I will investigate a bit more when I have time. I also plan to update the conan version we are using that is quite old already.

clanmills commented:
Please bring the source up-to-date.  The `unsatisfied` externals in the log concern **dlopen/dlclose/etc/**:

```
../lib/libexiv2.so.0.27.2.2: undefined reference to `dladdr'
../lib/libexiv2.so.0.27.2.2: undefined reference to `dlopen'
../lib/libexiv2.so.0.27.2.2: undefined reference to `dlclose'
../lib/libexiv2.so.0.27.2.2: undefined reference to `dlerror'
../lib/libexiv2.so.0.27.2.2: undefined reference to `dlsym'
```

The code in src/version.cpp which uses **dlopen/dlclose/etc** was changed (for v0.27.0, I think).  Instead of using the dynamic linker, Linux and Unix now read the process tables in /proc/this/that to determine the libraries which have been loaded at run-time.

On my build of exiv2 on Ubuntu 16.04, I can't see any indication that the linker needs **dlopen/dlclose/etc/** functions.

piponazo committed:
Conan: Updates to fix sudden linkage issues

We suddenly started to have some linking issues in the 0.27 branch after
the libcurl packages were updated in the conan-center repositories.
After some experimentation I took the following steps to fix up the
situation:

- Update conan to latest version
- Update of libcurl to the latest version available
- Use libcurl static libraries

Note that the change to use static libraries is just to make the
deployment step as easier as possible in the travis builds.

clanmills reviewed (dismissed):
We should remove all code relating `libdl` as dlopen/dlclose/etc are no longer used by our code.

piponazo commented:
> We should remove all code relating libdl as dlopen/dlclose/etc are no longer used by our code.

Sure, I'll do it in this branch too. I have been experimenting this morning a bit, and it seems that everything is compiling and linking properly finally (after updating conan, and the version of libcurl that we were using from the conan main repository.

I will squash the commits once I am done with the investigation.

piponazo commented:
Alright, I squashed the commits and this should be the final working version. If all the CI jobs pass I would merge this ASAP, since it is blocking the rest of PRs (all the builds will fail until we fix the situation)

D4N reviewed (approved):
Looks good, thanks for taking care of this!
